From 70b309645ce6bc264dac8b39973caa1a26a50ef5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?C=C3=A9dric=20VINCENT?= <cedric.vincent@st.com>
Date: Tue, 24 Mar 2020 17:17:06 +0100
Subject: [PATCH 1019/1695] drivers: add frame_buffer driver support
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Reverts commit cd12a61e11b5 ("TUI: remove frame buffer") to restore
frame buffer implementation that was deprecated due to lack of
maintenance.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 176bbcfc82d1d7e94f48456933bc2ced6057df72 drivers: add frame_buffer driver support]
Change-Id: I9eb9776178d33b0b75d79e8d71c2bf4f372e5879
Signed-off-by: Cédric VINCENT <cedric.vincent@st.com>
Signed-off-by: Yannick Fertre <yannick.fertre@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/391771
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/frame_buffer.c         | 47 ++++++++++++++++++++++++++
 core/drivers/sub.mk                 |  1 +
 core/include/display.h              | 52 +++++++++++++++++++++++++++++
 core/include/drivers/frame_buffer.h | 29 ++++++++++++++++
 4 files changed, 129 insertions(+)
 create mode 100644 core/drivers/frame_buffer.c
 create mode 100644 core/include/display.h
 create mode 100644 core/include/drivers/frame_buffer.h

diff --git a/core/drivers/frame_buffer.c b/core/drivers/frame_buffer.c
new file mode 100644
index 000000000..1ef82cf58
--- /dev/null
+++ b/core/drivers/frame_buffer.c
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2016, Linaro Limited
+ */
+
+#include <assert.h>
+#include <compiler.h>
+#include <drivers/frame_buffer.h>
+#include <util.h>
+
+size_t frame_buffer_get_image_size(struct frame_buffer *fb __unused,
+				   size_t width, size_t height)
+{
+	return width * height * sizeof(uint32_t);
+}
+
+void frame_buffer_clear(struct frame_buffer *fb, uint32_t color)
+{
+	uint32_t *base = 0;
+	size_t n = 0;
+
+	assert(fb && fb->base);
+	assert(IS_ALIGNED_WITH_TYPE(fb->base, uint32_t));
+	base = fb->base;
+
+	for (n = 0; n < fb->width * fb->height; n++)
+		base[n] = color;
+}
+
+void frame_buffer_set_image(struct frame_buffer *fb, size_t xpos, size_t ypos,
+			    size_t width, size_t height, const void *image)
+{
+	const uint32_t *img = image;
+	uint32_t *base = 0;
+	size_t x = 0;
+	size_t y = 0;
+
+	assert(fb && fb->base && img);
+	assert(IS_ALIGNED_WITH_TYPE(img, uint32_t));
+	assert(IS_ALIGNED_WITH_TYPE(fb->base, uint32_t));
+	base = fb->base;
+
+	for (y = 0; y < height && (y + ypos) < fb->height; y++)
+		for (x = 0; x < width && (x + xpos) < fb->width; x++)
+			base[x + xpos + (y + ypos) * fb->width] =
+				img[x + y * width];
+}
diff --git a/core/drivers/sub.mk b/core/drivers/sub.mk
index 11533597d..b88ee6810 100644
--- a/core/drivers/sub.mk
+++ b/core/drivers/sub.mk
@@ -4,6 +4,7 @@ srcs-$(CFG_CDNS_UART) += cdns_uart.c
 srcs-$(CFG_PL011) += pl011.c
 srcs-$(CFG_TZC400) += tzc400.c
 srcs-$(CFG_TZC380) += tzc380.c
+srcs-$(CFG_FRAME_BUFFER) += frame_buffer.c
 srcs-$(CFG_GIC) += gic.c
 srcs-$(CFG_CORE_HAFNIUM_INTC) += hfic.c
 srcs-$(CFG_PL061) += pl061_gpio.c
diff --git a/core/include/display.h b/core/include/display.h
new file mode 100644
index 000000000..ec0ad663a
--- /dev/null
+++ b/core/include/display.h
@@ -0,0 +1,52 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2016-2021, Linaro Limited
+ */
+
+#ifndef __DISPLAY_H
+#define __DISPLAY_H
+
+#include <drivers/frame_buffer.h>
+#include <tee_api_types.h>
+#include <sys/queue.h>
+
+struct disp_dev_list {
+	void *device;
+	TEE_Result (*device_init)(void *device);
+	TEE_Result (*device_final)(void *device);
+	TEE_Result (*device_activate)(void *device,
+				      const struct frame_buffer *fb,
+				      uint32_t x0, uint32_t y0);
+	TEE_Result (*device_get_display_size)(void *device, uint32_t *width,
+					      uint32_t *height);
+	SLIST_ENTRY(disp_dev_list) link;
+};
+
+#ifdef CFG_DISPLAY
+TEE_Result display_init(void);
+struct frame_buffer *display_get_frame_buffer(void);
+void display_final(void);
+void display_register_device(struct disp_dev_list *ddev);
+TEE_Result display_get_fb_addr_from_dtb(paddr_t *pa, size_t *size);
+#else
+static inline TEE_Result display_init(void)
+{ return TEE_ERROR_NOT_SUPPORTED; }
+
+static inline struct frame_buffer *display_get_frame_buffer(void)
+{ return NULL; }
+
+static inline void display_final(void)
+{ }
+
+static inline void display_register_device(struct disp_dev_list *ddev __unused)
+{ }
+
+static inline TEE_Result display_get_fb_addr_from_dtb(paddr_t *pa __unused,
+						      size_t *size __unused)
+{
+	return TEE_ERROR_NOT_SUPPORTED;
+}
+#endif /* CFG_DISPLAY */
+
+#endif /* __DISPLAY_H */
+
diff --git a/core/include/drivers/frame_buffer.h b/core/include/drivers/frame_buffer.h
new file mode 100644
index 000000000..4e504ea0d
--- /dev/null
+++ b/core/include/drivers/frame_buffer.h
@@ -0,0 +1,29 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2016-2021, Linaro Limited
+ */
+#ifndef __DRIVERS__FRAME_BUFFER_H
+#define __DRIVERS__FRAME_BUFFER_H
+
+#include <types_ext.h>
+
+enum frame_buffer_bpp {
+	FB_24BPP,
+};
+
+struct frame_buffer {
+	size_t width;
+	size_t height;
+	size_t width_dpi;
+	size_t height_dpi;
+	enum frame_buffer_bpp bpp;
+	void *base;
+};
+
+size_t frame_buffer_get_image_size(struct frame_buffer *fb, size_t width,
+				   size_t height);
+void frame_buffer_clear(struct frame_buffer *fb, uint32_t color);
+void frame_buffer_set_image(struct frame_buffer *fb, size_t xpos, size_t ypos,
+			    size_t width, size_t height, const void *image);
+
+#endif /*__DRIVERS__FRAME_BUFFER_H*/
-- 
2.47.2

